### DISCUSSION TITLE
libvips 8.15.1  thumbnail command on centos 7  is very slower than same version on ubuntu 20.04。

### DISCUSSION THREAD

**Discussion by stonebottle** [libvips/libvips#3870] [Q&A] [answered]
I have two virtual machine:
ENV 1：vmware , ubuntu 20.04 ,13600KF。 
      vipsthumbnail temp-17: done in 0.355s          
ENV 2：alibaba ecs(same as aws)， 16 core cpu，3.3GHZ
     vipsthumbnail temp-17: done in 1.12s     

thumbnail  command on env2 is very slower than on env1
why?

my services  heavily depend on imagemagick, I want to switch it to libvips.

details info:

 ### ENV2:

[XXX@iZbp10alfjwub4Z compare_vips_im]# bash run_vips.sh 
args cnt:0
args 1:
VIPS-INFO: 10:59:41.568: thumbnailing ./test_vips_jpg/n1.jpg
vips cache+: jpegload filename="./test_vips_jpg/n1.jpg" out=((VipsImage*) 0x13391b0) flags=((VipsForeignFlags) VIPS_FOREIGN_SEQUENTIAL) -
VIPS-INFO: 10:59:41.568: selected loader is VipsForeignLoadJpegFile
VIPS-INFO: 10:59:41.568: input size is 4458 x 6686
VIPS-INFO: 10:59:41.568: loading with factor 1 pre-shrink
vips cache : jpegload filename="./test_vips_jpg/n1.jpg" out=((VipsImage*) 0x1339340) shrink=1 flags=((VipsForeignFlags) VIPS_FOREIGN_SEQUENTIAL) access=((VipsAccess) VIPS_ACCESS_SEQUENTIAL) fail-on=((VipsFailOn) VIPS_FAIL_ON_NONE) -
VIPS-INFO: 10:59:41.568: pre-shrunk size is 4458 x 6686
VIPS-INFO: 10:59:41.568: converting to processing space srgb
vips cache+: colourspace in=((VipsImage*) 0x1339340) out=((VipsImage*) 0x13394d0) space=((VipsInterpretation) VIPS_INTERPRETATION_sRGB) -
vips cache : copy in=((VipsImage*) 0x13394d0) out=((VipsImage*) 0x13397f0) -
VIPS-INFO: 10:59:41.568: residual reducev by 0.299133
VIPS-INFO: 10:59:41.568: reducev: 21 point mask
vips cache : copy in=((VipsImage*) 0x13397f0) out=((VipsImage*) 0x1339b10) -
vips cache+: embed in=((VipsImage*) 0x1339b10) out=((VipsImage*) 0x1339ca0) x=0 y=10 width=4458 height=6707 extend=((VipsExtend) VIPS_EXTEND_COPY) -
VIPS-INFO: 10:59:41.569: reducev sequential line cache
vips cache+: linecache in=((VipsImage*) 0x1339e30) out=((VipsImage*) 0x13e81b0) tile-height=10 access=((VipsAccess) VIPS_ACCESS_SEQUENTIAL) -
vips cache+: sequential in=((VipsImage*) 0x1339e30) out=((VipsImage*) 0x13e8020) tile-height=10 -
vips cache+: reducev in=((VipsImage*) 0x13397f0) out=((VipsImage*) 0x1339980) vshrink=3.343000 kernel=((VipsKernel) VIPS_KERNEL_LANCZOS3) gap=2.000000 -
VIPS-INFO: 10:59:41.569: residual reduceh by 0.299133
VIPS-INFO: 10:59:41.569: reduceh: 21 point mask
vips cache : copy in=((VipsImage*) 0x1339980) out=((VipsImage*) 0x13e84d0) -
vips cache+: embed in=((VipsImage*) 0x13e84d0) out=((VipsImage*) 0x13e8660) x=10 y=0 width=4479 height=2000 extend=((VipsExtend) VIPS_EXTEND_COPY) -
vips cache+: reduceh in=((VipsImage*) 0x1339980) out=((VipsImage*) 0x13e8340) hshrink=3.343000 kernel=((VipsKernel) VIPS_KERNEL_LANCZOS3) gap=2.000000 -
vips cache+: resize in=((VipsImage*) 0x13394d0) out=((VipsImage*) 0x1339660) vscale=0.299133 scale=0.299133 -
vips cache : thumbnail filename="./test_vips_jpg/n1.jpg" out=((VipsImage*) 0x13e87f0) width=2000 height=2000 size=((VipsSize) VIPS_SIZE_BOTH) no-rotate=FALSE crop=((VipsInteresting) VIPS_INTERESTING_NONE) linear=FALSE import-profile=NULL export-profile=NULL intent=((VipsIntent) VIPS_INTENT_RELATIVE) -
VIPS-INFO: 10:59:41.569: thumbnailing ./test_vips_jpg/n1.jpg as ./test_vips_jpg/output1.jpg
vips cache+: cast in=((VipsImage*) 0x13e87f0) out=((VipsImage*) 0x13e8980) format=((VipsBandFormat) VIPS_FORMAT_UCHAR) -
vips cache : copy in=((VipsImage*) 0x13e8980) out=((VipsImage*) 0x13e8b10) -
vipsthumbnail temp-17: 1334 x 2000 pixels, 16 threads, 1334 x 16 tiles, 384 lines in buffer
vips cache+: linecache in=((VipsImage*) 0x7f50140461a0) out=((VipsImage*) 0x7f50140464c0) tile-height=8 access=((VipsAccess) VIPS_ACCESS_SEQUENTIAL) -
vips cache+: sequential in=((VipsImage*) 0x7f50140461a0) out=((VipsImage*) 0x7f5014046330) tile-height=8 -
vips cache+: extract_area input=((VipsImage*) 0x7f5014046330) out=((VipsImage*) 0x7f5014046650) left=0 top=0 width=4458 height=6686 -
VIPS-INFO: 10:59:42.684: threadpool completed with 4 workers
vipsthumbnail temp-17: done in 1.12s          
vips cache : jpegsave_target in=((VipsImage*) 0x13e87f0) target=((VipsTarget*) 0x1400000) -
vips_threadset_free: peak of 11 threads
memory: high-water mark 71.08 MB

**env1:**

xxx@xxx-virtual-machine:~/workspace/compare_vips_im$ bash run_vips.sh 
args cnt:0
args 1:
VIPS-INFO: 11:01:07.884: thumbnailing ./test_vips_jpg/n1.jpg
vips cache+: jpegload filename="./test_vips_jpg/n1.jpg" out=((VipsImage*) 0x55f4ea59f1b0) flags=((VipsForeignFlags) VIPS_FOREIGN_SEQUENTIAL) -
VIPS-INFO: 11:01:07.885: selected loader is VipsForeignLoadJpegFile
VIPS-INFO: 11:01:07.885: input size is 4458 x 6686
VIPS-INFO: 11:01:07.885: loading with factor 1 pre-shrink
vips cache : jpegload filename="./test_vips_jpg/n1.jpg" out=((VipsImage*) 0x55f4ea59f340) shrink=1 flags=((VipsForeignFlags) VIPS_FOREIGN_SEQUENTIAL) access=((VipsAccess) VIPS_ACCESS_SEQUENTIAL) fail-on=((VipsFailOn) VIPS_FAIL_ON_NONE) -
VIPS-INFO: 11:01:07.886: pre-shrunk size is 4458 x 6686
VIPS-INFO: 11:01:07.886: converting to processing space srgb
vips cache+: colourspace in=((VipsImage*) 0x55f4ea59f340) out=((VipsImage*) 0x55f4ea59f4d0) space=((VipsInterpretation) VIPS_INTERPRETATION_sRGB) -
vips cache : copy in=((VipsImage*) 0x55f4ea59f4d0) out=((VipsImage*) 0x55f4ea59f7f0) -
VIPS-INFO: 11:01:07.886: residual reducev by 0.299133
VIPS-INFO: 11:01:07.886: reducev: 21 point mask
vips cache : copy in=((VipsImage*) 0x55f4ea59f7f0) out=((VipsImage*) 0x55f4ea59fb10) -
vips cache+: embed in=((VipsImage*) 0x55f4ea59fb10) out=((VipsImage*) 0x55f4ea59fca0) x=0 y=10 width=4458 height=6707 extend=((VipsExtend) VIPS_EXTEND_COPY) -
VIPS-INFO: 11:01:07.887: reducev: using vector path
VIPS-INFO: 11:01:07.887: reducev sequential line cache
vips cache+: linecache in=((VipsImage*) 0x55f4ea59fe30) out=((VipsImage*) 0x55f4ea6791a0) tile-height=10 access=((VipsAccess) VIPS_ACCESS_SEQUENTIAL) -
vips cache+: sequential in=((VipsImage*) 0x55f4ea59fe30) out=((VipsImage*) 0x55f4ea679010) tile-height=10 -
vips cache+: reducev in=((VipsImage*) 0x55f4ea59f7f0) out=((VipsImage*) 0x55f4ea59f980) vshrink=3.343000 kernel=((VipsKernel) VIPS_KERNEL_LANCZOS3) gap=2.000000 -
VIPS-INFO: 11:01:07.887: residual reduceh by 0.299133
VIPS-INFO: 11:01:07.887: reduceh: 21 point mask
vips cache : copy in=((VipsImage*) 0x55f4ea59f980) out=((VipsImage*) 0x55f4ea6794c0) -
vips cache+: embed in=((VipsImage*) 0x55f4ea6794c0) out=((VipsImage*) 0x55f4ea679650) x=10 y=0 width=4479 height=2000 extend=((VipsExtend) VIPS_EXTEND_COPY) -
vips cache+: reduceh in=((VipsImage*) 0x55f4ea59f980) out=((VipsImage*) 0x55f4ea679330) hshrink=3.343000 kernel=((VipsKernel) VIPS_KERNEL_LANCZOS3) gap=2.000000 -
vips cache+: resize in=((VipsImage*) 0x55f4ea59f4d0) out=((VipsImage*) 0x55f4ea59f660) vscale=0.299133 scale=0.299133 -
vips cache : thumbnail filename="./test_vips_jpg/n1.jpg" out=((VipsImage*) 0x55f4ea6797e0) width=2000 height=2000 size=((VipsSize) VIPS_SIZE_BOTH) no-rotate=FALSE crop=((VipsInteresting) VIPS_INTERESTING_NONE) linear=FALSE import-profile=NULL export-profile=NULL intent=((VipsIntent) VIPS_INTENT_RELATIVE) -
VIPS-INFO: 11:01:07.887: thumbnailing ./test_vips_jpg/n1.jpg as ./test_vips_jpg/output1.jpg
vips cache+: cast in=((VipsImage*) 0x55f4ea6797e0) out=((VipsImage*) 0x55f4ea679970) format=((VipsBandFormat) VIPS_FORMAT_UCHAR) -
vips cache : copy in=((VipsImage*) 0x55f4ea679970) out=((VipsImage*) 0x55f4ea679b00) -
vipsthumbnail temp-17: 1334 x 2000 pixels, 16 threads, 1334 x 16 tiles, 384 lines in buffer
vips cache+: linecache in=((VipsImage*) 0x7f478c0551b0) out=((VipsImage*) 0x7f478c0554d0) tile-height=8 access=((VipsAccess) VIPS_ACCESS_SEQUENTIAL) -
vips cache+: sequential in=((VipsImage*) 0x7f478c0551b0) out=((VipsImage*) 0x7f478c055340) tile-height=8 -
vips cache+: extract_area input=((VipsImage*) 0x7f478c055340) out=((VipsImage*) 0x7f478c055660) left=0 top=0 width=4458 height=6686 -
vipsthumbnail temp-17: done in 0.355s          
vips cache : jpegsave_target in=((VipsImage*) 0x55f4ea6797e0) target=((VipsTarget*) 0x55f4ea6ca900) -
vips_threadset_free: peak of 11 threads
memory: high-water mark 71.94 MB

**Comment by stonebottle** [comment]
On ENV1 :    print a different info: " reducev: using vector path"
On ENV2：  no using vector path

why?  how to enable using the vector path ?

**Comment by stonebottle** [comment]
knee down , online waiting

**Comment by jcupitt** [comment, marked as answer]
Hi, libvips 8.15 uses highway for vector processing. You'll need to install that before building libvips.

**Reply by stonebottle** [reply]
ok, done.
recompile libvips with highway lib, the problem is resolved
